pinot-broker/src/main/java/com/linkedin/pinot/broker/queryquota/HitCounter.java

```diff
@@ -0,0 +1,58 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.broker.queryquota;
+
+public class HitCounter {
+
+  private static int TIME_WINDOW = 100;
+  private int _timeRangeInSecond;
+  private int _timeUnit;
+  private long[] _times;
+  private long[] _hits;
+
+  public HitCounter(int timeRangeInSecond) {
+    _timeRangeInSecond = timeRangeInSecond;
+    _timeUnit = _timeRangeInSecond * 1000 / TIME_WINDOW;
+    _times = new long[TIME_WINDOW];
+    _hits = new long[TIME_WINDOW];
+  }
+
+  public void hit(long timestamp) {
+    long numTimeUnits = timestamp / _timeUnit;
+    int index = (int)(numTimeUnits % TIME_WINDOW);
+    if (_times[index] == numTimeUnits) {
```

Comment:
Seems like you'd need a compare-and-swap op for multithreaded correctness here.

Reply:
Good catch! 👍 I'll use `AtomicInteger` in this case. Thx!

Reply:
Your `if` statement is dependent on all threads doing no stale reads and no parallel execution of the `if`, which is not the case.

Quick test for multithreaded correctness:

```
public class TestHitCounter {
  public static void main(String[] args) {
    int expectedHitCount = 30 * 200000;
    HitCounter hitCounter = new HitCounter(10);

    List<Thread> threadList = new ArrayList<>();


    for (int i = 0; i < 30; i++) {
      Thread thread = new Thread() {
        @Override
        public void run() {
          for (int j = 0; j < 200000; j++) {
            hitCounter.hit(System.currentTimeMillis());
          }
        }
      };
      thread.start();
      threadList.add(thread);
    }

    for (Thread thread : threadList) {
      Uninterruptibles.joinUninterruptibly(thread);
    }

    System.out
        .println("hitCounter.getHits(System.currentTimeMillis()) = " + hitCounter.getHits(System.currentTimeMillis()));
  }
}
```